File: test_resources/kotlin/struct_basic.kt
data class ViewData(
    val deltaTime: Float,
    val viewWidth: Float,
    val viewHeight: Float,
    val touchStartX: Float,
    val touchStartY: Float,
    val lastTouchX: Float,
    val lastTouchY: Float,
    val touchX: Float,
    val touchY: Float,
) {
    companion object {
        fun createDefault(): ViewData = ViewData(
            deltaTime = 0f,
            viewWidth = 0f,
            viewHeight = 0f,
            touchStartX = 0f,
            touchStartY = 0f,
            lastTouchX = 0f,
            lastTouchY = 0f,
            touchX = 0f,
            touchY = 0f,
        )

        fun readFromBuffers(reader: Long): ViewData {
            val deltaTime = Float.readFromBuffers(reader)
            val viewWidth = Float.readFromBuffers(reader)
            val viewHeight = Float.readFromBuffers(reader)
            val touchStartX = Float.readFromBuffers(reader)
            val touchStartY = Float.readFromBuffers(reader)
            val lastTouchX = Float.readFromBuffers(reader)
            val lastTouchY = Float.readFromBuffers(reader)
            val touchX = Float.readFromBuffers(reader)
            val touchY = Float.readFromBuffers(reader)

            return ViewData(
                deltaTime = deltaTime,
                viewWidth = viewWidth,
                viewHeight = viewHeight,
                touchStartX = touchStartX,
                touchStartY = touchStartY,
                lastTouchX = lastTouchX,
                lastTouchY = lastTouchY,
                touchX = touchX,
                touchY = touchY,
            )
        }

        fun skipInBuffers(reader: Long, count: Int) {
            for (i in 0..<count) {
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
                Float.readFromBuffers(reader)
            }
        }
    }

    fun writeToBuffers(writer: Long) {
        deltaTime.writeToBuffers(writer)
        viewWidth.writeToBuffers(writer)
        viewHeight.writeToBuffers(writer)
        touchStartX.writeToBuffers(writer)
        touchStartY.writeToBuffers(writer)
        lastTouchX.writeToBuffers(writer)
        lastTouchY.writeToBuffers(writer)
        touchX.writeToBuffers(writer)
        touchY.writeToBuffers(writer)
    }
}

File: test_resources/dart/struct_basic.dart
class ViewData {
  const ViewData({
    required this.deltaTime,
    required this.viewWidth,
    required this.viewHeight,
    required this.touchStartX,
    required this.touchStartY,
    required this.lastTouchX,
    required this.lastTouchY,
    required this.touchX,
    required this.touchY,
  });

  const ViewData.createDefault()
      : deltaTime = 0.0,
        viewWidth = 0.0,
        viewHeight = 0.0,
        touchStartX = 0.0,
        touchStartY = 0.0,
        lastTouchX = 0.0,
        lastTouchY = 0.0,
        touchX = 0.0,
        touchY = 0.0;

  final double deltaTime;
  final double viewWidth;
  final double viewHeight;
  final double touchStartX;
  final double touchStartY;
  final double lastTouchX;
  final double lastTouchY;
  final double touchX;
  final double touchY;
}

class ViewDataBuffersFactory implements BuffersFactory<ViewData> {
  const ViewDataBuffersFactory();

  @override
  ViewData createDefault() => const ViewData.createDefault();
}
